README.md
# CSSP

## Summary

CSS with Padding. This project implements a plugin for the [RequireJS](http://requirejs.org/) project that allows modules to list CSS files as dependencies. This allows RequireJS to wait until specific style rules are known to be in effect before it releases the JavaScript module to be evaluated. This would be useful if, for example, if your JavaScript module required certain CSS rules, defined in a separate file, to be applied to the DOM before it could be initialized.

## What CSSP Is

The technique of polling for a known style rule to become applied, as a way of determining when an external stylesheet is loaded, has been successfully used in production-quality libraries like [BBC Glow](http://github.com/glow/glow1/blob/master/src/widgets/widgets.js#L33) for years. This idea has recently been proposed as a wider standard by [Pete Otaqui](http://otaqui.com/blog/890/cssp-loading-css-with-javascript-and-getting-an-onload-callback/). The basic idea is that the tool repeatedly checks a known DOM node for a known CSS property and value. When that node is found to have that particular style applied, we can then assume that the file that defines that rule has been successfully loaded by the browser.

## Usage

In the initial release, our implementation assumes that the CSS rule to watch for will be like this:

	#someID { position:relative; z-index:12345; }

That rule is the P (padding) in CSSP, because you would need to add this "signal" to the end of your CSS file in order for the CSSP plugin to detect that the stylesheet was in effect. Having that, and the CSSP plugin, would allow you to write a RequireJS dependency like so:

	define(
		'someModule',
		['someOtherModule', 'cssp!someStyles.css?someID'],
		
		function(someOtherModule, linkNode) {
			// we can be certain the someStyles.css
			// CSS rules are now in effect
			
			// the linkNode argument contains a reference to the DOM Node
			// associated with the injected LINK element for the stylesheet
			
			// module code goes here
		}
	);

Or simply...

	require({
		['cssp!someStyles.css?someID'],
		function() {
			// someStyles.css rules are in effect now
			
			// module code goes here
		}
	);

Notice it is not necessary that there actually be any element on the page with given signal ID. The plugin will automatically create one, watch it, and then remove it when the dependency is eventually fulfilled.

## Configuring Paths for cssp! IDs

The RequireJS `require()` method allows you to configure paths for you JavaScript modules by associating a moduleID with a specific file path. If you use a moduleID that starts with "cssp!" the associated path will be used to include the `.css` file. For example:

    require({
        baseUrl: 'base/dir/',
	    paths: {
            'cssp!sparkle/unicorns': 'css/my-magic.css',
            'cssp!sparkle/elves':    'css/my-magic.css'
        }
    });
    
    // later...
    
    require({
		['cssp!sparkle/elves?someID'],
		function() {
			// base/dir/css/my-magic.css rules are in effect now
			
			// module code goes here
		}
	);

Notice that by using this technique you can associate multiple cssp! moduleIDs with a single `.css` file. Also, as an efficiency, the CSS file inclusion will only ever happen once per filepath, regardless of how many different moduleIDs are associated with that filepath.

### Using the Default Signal ID

In the above examples the user provided a signal ID to the the CSSP plugin by appending it after a `?` character. It is possible to allow the CSSP plugin to derive and use a default signal ID however. The ID will be calculated by taking the given css ID and replacing all non-alphanumeric characters with a dash. So, for example, the following cssp requires are identical:

    require({
		['cssp!sparkle/elves?cssp-sparkle-elves'],
		function() {
		}
	);
	
	// is the same as...
	
	require({
		['cssp!sparkle/elves'],
		function() {
		}
	);


CHANGES
0.4.1
 - Fixed a bug - removed sniffing for IE and over-writing of the global getElementById method since this was causing errors in IE9
 - Optimized loop in the look function by reducing DOM queries for an element

0.4.0
  - Updated to use plugin API for RequireJS 0.24.0

0.3.0
  - Callback for a cssp dependency now receives a ref to the injected link DOM
    node as its argument.

0.2.0
  - Added support for cssp! module ID's in the requireJS paths configuration.
  - Added support for tracking css file injections, so same stylesheet is only
  ever loaded once.

0.0.1
  - Initial release: Support dependency on CSS files in RequireJS.


test/tests/basic.js
(function() {
    
    require(
        {
            baseUrl: 'mock'
        }
    );
        
    function setup() {
        saveTestElement();
    }
     
    function teardown() {
        restoreTestElement();
    }
     
    asyncTest('CSSP can be used when no test element exists on the page.', function() {
        expect(3);
        setup();
        
        var csspTestEl = document.getElementById('addedByCssp');
        equals(csspTestEl, null, 'There should be no test element on the page initially.');
        
        require(
            ['cssp!test1.css?test1'],
            function() {
                ok(true, 'The css dependency should be fulfilled.');
                
                csspTestEl = document.getElementById('addedByCssp');
                equals(csspTestEl, null, 'There should still be no test element on the page after the css is detected.');
            
                teardown();
                start(); 
            }
        );
    });
    
    asyncTest('CSSP plugin works across multiple require calls.', function () {
        expect(2);
        
        require(
            ['cssp!test1.css?test1'],
            function() {
                ok(true, 'The css dependency should be fulfilled.');
            }
        );
        
        require(
            ['cssp!test1.css?test1'],
            function() {
                ok(true, 'The css dependency should be fulfilled in a second separate call to require().');
            }
        );
        
        setTimeout(function() {
            teardown();
            start();
        }, 2000);
        
        
    });
    
     
     asyncTest('CSSP can be used when a test element already exists on the page.', function() {
        expect(3);
        setup();
        
        var csspTestEl = document.getElementById('test2');
        ok(csspTestEl !== null, 'There should be a test element on the page initially.');
        
        require(['cssp!test2.css?test2'],
            function() {
                ok(true, 'The css dependency should be fulfilled.');
                
                csspTestEl = document.getElementById('test2');
                ok(csspTestEl !== null, 'There should still be a test element on the page after the css is detected.');
            
                teardown();
                start();
            }
        );
    });
     
    asyncTest('CSSP callback receives a ref to the injected link DOM node.', function() {
        expect(3);
        setup();
    
        require(
            {
                baseUrl: 'mock'
            },
            ['cssp!test3.css?test3'],
            function(linkNode) {
                equals(typeof linkNode, 'object', 'The given linkNode is defined.');
                
                equals(linkNode.nodeName.toUpperCase(), 'LINK', 'The given node object has a name of LINK.');
                ok(~ linkNode.href.indexOf('test3.css'), 'The given linkNode has the correct href property.');
            
                teardown();
                start();
            }
        );
    });

})();

test/mock/js/my-magic.js
define('sparkle/unicorns',
    { mock: true}
);

src/cssp.js
/**
 * @fileoverview Add the ability to depend on CSS files to RequireJS.
 * @copyright 2010 British Broadcasting Corporation
 * @author Richard Hodgson <richard.hodgson@bbc.co.uk>
 * @author Michael Mathews <michael.mathews@bbc.co.uk>
 * @see http://otaqui.com/blog/890/cssp-loading-css-with-javascript-and-getting-an-onload-callback/
 */

(function () {
    var magicNumber = 12345, // a zindex must be set to this to signal CSS has been applied
        csspQueue = [], // waiting for these: like [ [elemId, callback], [elemId, callback], ... ]
        intervalId,
        cacheTrack = {}; // what css files have already been added to this page?

    /*
     * Fix bugginess in IE that can cause the wrong element to be returned when
     * a DOM node has the same `name` as another DOM node's `id`.
     * @see http://www.sixteensmallstones.org/ie-javascript-bugs-overriding-internet-explorers-documentgetelementbyid-to-be-w3c-compliant-exposes-an-additional-bug-in-getattributes
     * @see http://www.w3.org/TR/2000/WD-DOM-Level-1-20000929/level-one-html.html#ID-36113835
     */
    function getElement(elemId){

        var elem = document.getElementById(elemId);

        if (elem) {
            if( elem.attributes['id'].value === elemId ) { //make sure that it is a valid match on id
                return elem;
            }
            else { //otherwise find the correct element
                var elements = document.all[elemId];
                for (var i = 1, leni = elements.length; i < leni; i++) { // returns the first match
                    if( elements[i].attributes['id'].value === elemId ) {
                        return elements[i]; 
                    }
                }
            }
        }
        return null; // as per the W3 spec
    }

    /**
     * Given an element's id, returns the current z-index of the element
     * @private
     * @param {string} elem The DOM element to check.
     * @returns {number}
     */
    function getZindex(elem) {
        var zIndex;

        if (elem) {
            // the w3c way, uses hyphenated property names
            if (window.getComputedStyle) {
                zIndex = document.defaultView.getComputedStyle(elem, null).getPropertyValue('z-index');
            }
            // the IE way
            else if (elem.currentStyle) {
                zIndex = elem.currentStyle['zIndex'];
            }
            zIndex = +zIndex;

            return zIndex;
        }
    }

    /**
     * Loop through the queue, looking for any node id's that match the magic number.
     * @private
     */
    function look() {
        var cssp,
            elem;
            
        if (!document.body) {
            return;
        }

        for (var i = 0; i < csspQueue.length; i++) { // length may change during loop
            cssp = csspQueue[i];

            /**
             The following is for optimization.
             Query the DOM only once and cache the DOM element
             in the cssp[0] item
            */
            if(typeof cssp[0] === 'string'){
                elem = getElement(cssp[0]);
                // is there a test element to look for?
                if (!elem) {
                    elem = addTestElement(cssp[0]);
                }

                cssp[0] = elem;
            }
            else{
                elem = cssp[0];
            }

            if ( getZindex(elem) === magicNumber ) {
                cssp[1](); // fire callback
                csspQueue.splice(i--, 1); // remove this item from the queue, decrement i
                
                if (csspQueue.length === 0) { // can stop looking now
                    clearInterval(intervalId);
                }
            }
        }
    }

    /**
     * If there is no test element with the given ID, we add one ourselves.
     * @private
     */
    function addTestElement(elemId) {
        var elem = document.createElement('div');
        elem.id = elemId;
        elem.className = 'addedByCssp';
        elem.setAttribute('style', 'position: absolute; width: 1px; height: 1px; top: -1px; left: -1px;');
        
        document.body.insertBefore(elem, document.body.firstChild);
        
        return elem;
    }

    define(
        'cssp',
        {
        
        /**
         * Called when a dependency needs to be loaded.
         * @param {string} name Like some/url/foo.css?elementId
         * @param {Object} parentRequire reference to the parent context's require.
         * @param {function} load To call when the dependency is resolved.
         * @param {Object} config The parent context's configuration.
         */
        load: function (name, parentRequire, load, config) {

            var splitAt = name.lastIndexOf('?'),
                cssUrl = (splitAt > 0? name.substring(0, splitAt) : name),
                cssElementId = (splitAt > 0? name.substring(splitAt + 1, name.length) : 'cssp-' + name.replace(/[^a-z0-9_]/gi, '-') ),
                moduleId = 'cssp!' + cssUrl,
                head = document.getElementsByTagName('head')[0],
                cssLinkElement = document.createElement('link');

            // is there a path defined for this css?
            cssUrl = config.paths['cssp!'+cssUrl]? config.paths['cssp!'+cssUrl] : cssUrl;

            if (!cssUrl) {
                throw('CSS URL is required.');
            }
            
            // apply effects like the 'basepath' being set on this requirejs configuration
            cssUrl = parentRequire.toUrl(cssUrl);

            if ( cacheTrack[cssUrl] ) {   // already included a link to this css
                load( cacheTrack[cssUrl] ); // pass back the cached CSS Link element
                return;
            }
            cacheTrack[cssUrl] = cssLinkElement;

            // add this item to the queue
            csspQueue.push([cssElementId, function() { // create callback function
                // remove test element if it is one we added
                var elem = getElement(cssElementId);
                if (elem.className === 'addedByCssp') { // only remove test elements that we added
                    document.body.removeChild(elem);
                }
                
                load(cssLinkElement);
            }]);
            
            if (csspQueue.length === 1) {
                intervalId = setInterval(look, 250); // start looking now
            }

            cssLinkElement.type = 'text/css';
            cssLinkElement.rel  = 'stylesheet';
            cssLinkElement.href = cssUrl;
            head.appendChild(cssLinkElement); // order matters
        }
    });
}());
